File: darwin/Classes/Logger.swift
class Logger {
  // for singleton
  static let shared = Logger()
  private init() {}

  enum Level: Int {
    case debug
    case info
    case warning
    case error
    case none

    init(name: String) {
      switch name {
      case "debug": self = .debug
      case "info": self = .info
      case "warning": self = .warning
      case "error": self = .error
      default: self = .none
      }
    }
  }

  var level: Level = .none

  private static let pluginName: String = "UnifiedAppleVisionPlugin"

  static func _log(_ level: Level, _ message: String, _ symbol: String?) {
    if shared.level.rawValue < level.rawValue {
      return
    }
    var prefix = Logger.pluginName
    if let s = symbol {
      prefix += ">\(s)"
    }
    print("[\(prefix)] \(message)")
  }

  static func debug(_ message: String, _ symbol: String? = nil) {
    _log(.debug, message, symbol)
  }

  static func info(_ message: String, _ symbol: String? = nil) {
    _log(.info, message, symbol)
  }

  static func warning(_ message: String, _ symbol: String? = nil) {
    _log(.warning, message, symbol)
  }

  static func error(_ message: String, _ symbol: String? = nil) {
    _log(.error, message, symbol)
  }
}

File: lib/src/utility/logger.dart
import 'package:flutter/foundation.dart';
import 'package:unified_apple_vision/src/enum/log_level.dart';
import 'package:unified_apple_vision/src/extension/optional.dart';

class Logger {
  VisionLogLevel threshold = VisionLogLevel.debug;

  void log({
    required VisionLogLevel level,
    required LogSide side,
    required String msg,
    String? symbol,
  }) {
    if (level.index >= threshold.index) {
      debugPrint('${level.emoji} '
          '[unified_apple_vision${side.emoji}${symbol.maybe((s) => '>$s') ?? ''}] '
          '$msg');
    }
  }

  void d(String s, [String? symbol]) => log(
        level: VisionLogLevel.debug,
        side: LogSide.dart,
        msg: s,
        symbol: symbol,
      );

  void i(String s, [String? symbol]) => log(
        level: VisionLogLevel.info,
        side: LogSide.dart,
        msg: s,
        symbol: symbol,
      );

  void w(String s, [String? symbol]) => log(
        level: VisionLogLevel.warning,
        side: LogSide.dart,
        msg: s,
        symbol: symbol,
      );

  void e(String s, {Object? error, StackTrace? stackTrace, String? symbol}) {
    log(
      level: VisionLogLevel.error,
      side: LogSide.dart,
      msg: s,
      symbol: symbol,
    );
    if (error != null) {
      debugPrint(error.toString());
    }
    if (stackTrace != null) {
      debugPrint(stackTrace.toString());
    }
  }
}

extension LogLevelEx on VisionLogLevel {
  String get emoji => switch (this) {
        VisionLogLevel.debug => '',
        VisionLogLevel.info => 'ℹ️',
        VisionLogLevel.warning => '🟡',
        VisionLogLevel.error => '🚫',
        VisionLogLevel.none => '',
      };
}

enum LogSide {
  swift('🍏'),
  dart('🎯'),
  ;

  final String emoji;
  const LogSide(this.emoji);
}
